Bio/Align/substitution_matrices/__init__.py

```diff
@@ -457,15 +457,12 @@ def read(handle, dtype=float):
     """Parse the file and return an Array object."""
     try:
         fp = open(handle)
-        lines = fp.readlines()
     except TypeError:
         fp = handle
-        try:
-            lines = fp.readlines()
-        except Exception as e:
-            raise e from None
-        finally:
-            fp.close()
+
+    with fp:
```

Comment:
this does not seem to work if `fp` is a file name.

Reply:
I think it shouldn't be possible for `fp` to be a filename here, unless I'm mistaken. if `handle` is a string or path, `open(handle)` will not raise `TypeError`, so we should never hit `fp = handle` in that case. By the time we get to `with fp`, `handle` should always be a file handle as far as I can tell -- both in my revision and in the original.

Reply:
It looks like all of the existing tests for `substitution_matrices.read` pass `handle` as a string https://github.com/biopython/biopython/blob/64ef618a087e555561b4bb47a73c4bfb742a8f36/Tests/test_align_substitution_matrices.py#L112

@byerlyj you could add a unittest that passes an open filehandle to make sure both paths are covered?

Reply:
you got it! 👍